src/pybind/mgr/cephadm/module.py

```diff
@@ -518,6 +533,32 @@ def serve(self) -> None:
             self._serve_sleep()
         self.log.debug("serve exit")
 
+    def convert_tags_to_repo_digest(self):
+        if not self.use_repo_digest:
+            return
+        settings = self.upgrade.get_distinct_container_image_settings()
+        digests: Dict[str, ContainerInspectInfo] = {}
+        for container_image_ref in set(settings.values()):
+            if not is_repo_digest(container_image_ref):
+                image_info = self._get_container_image_info(container_image_ref)
+                if image_info.repo_digest:
+                    assert is_repo_digest(image_info.repo_digest), image_info
+                digests[container_image_ref] = image_info
+
+        for entity, container_image_ref in settings.items():
+            if not is_repo_digest(container_image_ref):
+                image_info = digests[container_image_ref]
+                if image_info.repo_digest:
+                    self.set_container_image(entity, image_info.repo_digest)
+
```

Comment:
minor nit: this is difficult to read, maybe a few comments about what each loop does would help .. 

basically we are building a list of digests and _only_ setting the container image iff the config value is a label (and not an existing digest).